### ISSUE TITLE
Models loaded from files do not perform as before saving

### ISSUE THREAD

**Issue by dmitra79** [tensorflow/tensorflow#43230]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04.2 LTS (GNU/Linux 4.15.0-45-generic x86_64)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device:
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): v2.3.0-rc2-23-gb36436b087 2.3.0
- Python version: 3.7.5 
- Bazel version (if compiling from source): NA
- GCC/Compiler version (if compiling from source): NA
- CUDA/cuDNN version:  CUDA Version: 10.1 
- GPU model and memory:   NVIDIA-SMI 430.50       Driver Version: 430.50      

You can collect some of this information using our environment capture [script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with:
1. TF 1.0: `python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"`
2. TF 2.0: `python -c "import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I created a semi-supervised GAN model, based on: https://machinelearningmastery.com/semi-supervised-generative-adversarial-network/. It works with MNIST data.
During the training the model reaches >90% accuracy, and gets saved in a file.
However, when I try to load the saved model and run it on the same data, it performs at random (~10% accuracy).
I tried both SavedModel and H5 formats, but I get the same results. 

**Describe the expected behavior**
I would have expected the model loaded from the file to result in same performance as the model before saving.

**Standalone code to reproduce the issue**

Collab notebook for training: https://colab.research.google.com/drive/1-FnGgEhCPwEjyQHiSWLuwTtqRVb-frpT?usp=sharing
Collab notebook applying the model: https://colab.research.google.com/drive/1JDeMO1e3Z7RVbdW-QEC2V0yWbRtSz0jx?usp=sharing 
(Note: the code is meant to run on a machine, not in Collab)

**Comment by Saduf2019** [comment]
@dmitra79 
I ran the colab shared and face a different error, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/e461d52e8e3a9eea05cc270b312ab2da/untitled413.ipynb).

**Comment by dmitra79** [comment]
@Saduf2019 sorry, the code was run on my machine and I just copied it to Colab. The particular error is due to 'res/' directory not being in place. I've modified the notebooks to store in local folder directly. Btw, it should run faster with GPU.

**Comment by Saduf2019** [comment]
@dmitra79 
I ran the code sheared in colab and did not face any errors, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/8276d5cfd33ae02b3adf9745632c13dc/untitled414.ipynb). 
If this does not replicate your problem, please share a colab gist with issue reported.

**Comment by dmitra79** [comment]
@Saduf2019, there is no error thrown. The problem is that saving/loading a model results in a very different accuracy.
I updated the scenario so that a single notebook has training and loading/applying of the model: https://colab.research.google.com/drive/1-FnGgEhCPwEjyQHiSWLuwTtqRVb-frpT?usp=sharing

See the end of the output:
```
>2999, c[0.005,100], d[0.731,0.631], g[1.269]
>3000, c[0.007,100], d[0.714,0.621], g[1.113]
Classifier Accuracy: 93.730%
>Saved: res/generated_plot_3000.png, res/g_model_3000.h5, res/c_model_3000.h5 and res/c_model_3000_sm
Final Accuracy: 93.730%
WARNING:tensorflow:Error in loading the saved optimizer state. As a result, your model is starting with a freshly initialized optimizer.
WARNING:tensorflow:Error in loading the saved optimizer state. As a result, your model is starting with a freshly initialized optimizer.
Final Accuracy: 9.687%
```

**Comment by k-w-w** [comment]
Hi, thanks for the report! This has been fixed in TF 2.4.
